core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -41,15 +41,19 @@ import scala.collection.{Seq, Set, mutable}
  *  A cache for the state (e.g., current leader) of each partition. This cache is updated through
  *  UpdateMetadataRequest from the controller. Every broker maintains the same cache, asynchronously.
  */
-class MetadataCache(brokerId: Int) extends Logging {
+class MetadataCache(brokerId: Int, localClusterId: String = "defaultClusterId", federationEnabled: Boolean = false) extends Logging {
 
   private val partitionMetadataLock = new ReentrantReadWriteLock()
   //this is the cache state. every MetadataSnapshot instance is immutable, and updates (performed under a lock)
   //replace the value with a completely new one. this means reads (which are not under any lock) need to grab
   //the value of this var (into a val) ONCE and retain that read copy for the duration of their operation.
   //multiple reads of this value risk getting different snapshots.
-  @volatile private var metadataSnapshot: MetadataSnapshot = MetadataSnapshot(partitionStates = mutable.AnyRefMap.empty,
-    controllerId = None, aliveBrokers = mutable.LongMap.empty, aliveNodes = mutable.LongMap.empty)
+  @volatile private var metadataSnapshot: MetadataSnapshot = if (federationEnabled)
```

Comment:
the checking of federationEnabled is performed a handful of times in this class. Do you think it makes sense to have a subclass of MetadataCache to host the federation-specific logic?

Reply:
That's an excellent point--I had lost track of how many of those updateMetadata() ended up with thanks to the four "alive" variables.  I'll do that.